src/rgw/rgw_gc.cc

```diff
@@ -94,114 +94,121 @@ int RGWGC::send_chain(cls_rgw_obj_chain& chain, const string& tag, bool sync)
-    c->wait_for_safe();
-    ret = c->get_return_value();
+
+    // this tag may still be present in omap, so remove it once the cls_rgw_gc
+    // enqueue succeeds
+    cls_rgw_gc_remove(op, {tag});
+
+    auto c = librados::Rados::aio_create_completion(nullptr, nullptr, nullptr);
+    int ret = store->gc_aio_operate(obj_names[i], c, &op);
     c->release();
-    if ((ret < 0 && ret != -ENOENT)) {
-      return ret;
-    }
-    if (! transitioned_objects_cache[i]) {
-      //Successfully deferred
-      if (ret == 0) {
-        //Enqueue in queue
-        cls_rgw_gc_queue_enqueue(op, cct->_conf->rgw_gc_obj_min_wait, info);
-        //Remove tag from omap
-        vector<string> tags;
-        tags.emplace_back(tag);
-        cls_rgw_gc_remove(op, tags);
-      }
-      //If not found in omap, then the tag must be in queue
-      if (ret == -ENOENT) {
-        cls_rgw_gc_queue_defer_entry(op, cct->_conf->rgw_gc_obj_min_wait, info);
-      }
-      ret = store->gc_aio_operate(obj_names[i], &op, &c);
-      c->wait_for_safe();
-      ret = c->get_return_value();
-      c->release();
-      return ret;
-    }
-    if (transitioned_objects_cache[i]) {
-      return ret;
-    }
+    return ret;
   }
-  return 0;
+
+  // if we haven't seen the transition yet, write the defer to omap with cls_rgw
+  ObjectWriteOperation op;
+
+  // assert that we haven't initialized cls_rgw_gc queue. this prevents us
+  // from writing new entries to omap after the transition
+  obj_version objv; // objv.ver = 0
+  cls_version_check(op, objv, VER_COND_EQ);
+
+  cls_rgw_gc_defer_entry(op, cct->_conf->rgw_gc_obj_min_wait, tag);
+
+  // prepare a callback to detect the transition via ECANCELED from cls_version_check()
+  auto state = std::make_unique<defer_chain_state>();
+  state->gc = this;
+  state->info.chain = chain;
+  state->info.tag = tag;
+  state->completion = librados::Rados::aio_create_completion(
+      state.get(), async_defer_callback, nullptr);
+
+  int ret = store->gc_aio_operate(obj_names[i], state->completion, &op);
+  if (ret == 0) {
+    state.release(); // release ownership until async_defer_callback()
+  }
+  return ret;
 }
 
 int RGWGC::remove(int index, const std::vector<string>& tags, AioCompletion **pc)
 {
   ObjectWriteOperation op;
   cls_rgw_gc_remove(op, tags);
-  return store->gc_aio_operate(obj_names[index], &op, pc);
+
+  auto c = librados::Rados::aio_create_completion(nullptr, nullptr, nullptr);
+  int ret = store->gc_aio_operate(obj_names[index], c, &op);
+  if (ret < 0) {
+    c->release();
+  } else {
+    *pc = c;
+  }
+  return ret;
 }
 
 int RGWGC::remove(int index, int num_entries, librados::AioCompletion **pc)
 {
   ObjectWriteOperation op;
   cls_rgw_gc_queue_remove_entries(op, num_entries);
-  return store->gc_aio_operate(obj_names[index], &op, pc);
+
+  auto c = librados::Rados::aio_create_completion(nullptr, nullptr, nullptr);
+  int ret = store->gc_aio_operate(obj_names[index], c, &op);
+  if (ret < 0) {
```

Comment:
@cbodley : Can you please explain what this part of code (c->release()) is doing here.

Reply:
discussed offline, but to summarize: when a callback is given, we need to keep the AioCompletion around so the callback can access its return code, ie. completion->get_return_value(). the other cases without a callback can release the completion immediately after submission to gc_aio_operate()

Reply:
Thanks for the explanation!